In [1]:
!wget https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Training%20Data.csv
!wget https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Training%20Data%20Target.csv
!wget https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Test%20Data.csv
!wget https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Sample%20Submission.csv

--2021-07-19 10:57:44--  https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Training%20Data.csv
Resolving lead-files.nyc3.cdn.digitaloceanspaces.com (lead-files.nyc3.cdn.digitaloceanspaces.com)... 205.185.216.42, 205.185.216.10
Connecting to lead-files.nyc3.cdn.digitaloceanspaces.com (lead-files.nyc3.cdn.digitaloceanspaces.com)|205.185.216.42|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5835092 (5.6M) [text/csv]
Saving to: ‘Training Data.csv’

Training Data.csv   100%[===================>]   5.56M  --.-KB/s    in 0.1s    

2021-07-19 10:57:44 (58.3 MB/s) - ‘Training Data.csv’ saved [5835092/5835092]

--2021-07-19 10:57:44--  https://lead-files.nyc3.cdn.digitaloceanspaces.com/data/hackathon-2/Training%20Data%20Target.csv
Resolving lead-files.nyc3.cdn.digitaloceanspaces.com (lead-files.nyc3.cdn.digitaloceanspaces.com)... 205.185.216.42, 205.185.216.10
Connecting to lead-files.nyc3.cdn.digitaloceanspaces.com (lead-files.nyc3.cdn.digitaloceansp

In [2]:
import pandas as pd
import numpy as np

train_df = pd.read_csv("/content/Training Data.csv")
train_target_df = pd.read_csv("/content/Training Data Target.csv")

train_df.head(10)

user_id      aov    category
0    37327  29128.0      Phones
1    37327    354.0     Fashion
2    37327    460.0  Home Decor
3    37327    761.0     Fashion
4    41260  16658.0      Phones
5    41260    263.0  Home Decor
6    41260    988.0     Fashion
7    19360  10214.0      Phones
8    19360  16761.0      Phones
9    19360   6382.0      Phones

In [3]:
train_target_df.tail(15)

user_id   aov    category
13230    48806 -9999     Fitness
13231    48858 -9999     Fashion
13232    48925 -9999     Fashion
13233    48987 -9999  Home Decor
13234    49025 -9999     Fitness
13235    49084 -9999     Fashion
13236    49192 -9999     Fitness
13237    49200 -9999     Fashion
13238    49249 -9999     Fitness
13239    49292 -9999     Fitness
13240    49383 -9999     Fashion
13241    49590 -9999  Home Decor
13242    49799 -9999     Fitness
13243    49900 -9999     Fitness
13244    49912 -9999     Fashion

In [4]:
len(train_target_df.user_id.unique())

13245

In [5]:
len(train_target_df)

13245

In [6]:
train_target_df[train_target_df.user_id == 49025]

user_id   aov category
13234    49025 -9999  Fitness

In [7]:
train_df[train_df.user_id == 49025]

user_id      aov           category
163708    49025  18199.0  Consumer Durables
163709    49025  12847.0             Phones
163710    49025    373.0               Toys
163711    49025    696.0            Fashion
163712    49025   9589.0             Phones
163713    49025    667.0            Fashion
163714    49025   1059.0          Groceries
163715    49025    272.0         Home Decor
163716    49025    408.0         Home Decor
163717    49025   2625.0          Groceries
163718    49025    498.0            Fashion
163719    49025    919.0               Toys
163720    49025   7721.0            Fitness
255364    49025    999.0            Fashion

In [8]:
len(train_df.user_id.unique())

29972

In [9]:
train_target_df.iloc[0].category

'Phones'

In [10]:
train_df.category.value_counts()

Phones                       66341
Fashion                      60327
Home Decor                   29703
Toys                         21106
Groceries                    12810
Back to School               12566
Gaming                       11937
Consumer Durables            11731
Books                         6010
Kitchen cleaning Supplies     3205
Beauty Products               3016
Ereaders                      3012
Board Games                   3001
Painiting Supplies            2985
Pet Supplies                  2955
Fitness                       2945
Laptops                       1888
TVs                           1869
Name: category, dtype: int64

In [11]:
train_target_df.category.value_counts()

Phones                       7124
Fashion                      2195
Pet Supplies                  910
Ereaders                      887
Fitness                       874
TVs                           426
Laptops                       308
Home Decor                    238
Consumer Durables             186
Toys                           58
Back to School                 19
Gaming                         10
Groceries                       8
Kitchen cleaning Supplies       2
Name: category, dtype: int64

In [12]:
test_df = pd.read_csv('/content/Test Data.csv')
test_df.category.value_counts()

Phones                       44627
Fashion                      40082
Home Decor                   19911
Toys                         14096
Groceries                     8559
Back to School                8405
Gaming                        8109
Consumer Durables             7984
Books                         3968
Ereaders                      2163
Pet Supplies                  2059
Beauty Products               2055
Kitchen cleaning Supplies     2042
Painiting Supplies            1987
Board Games                   1981
Fitness                       1961
Laptops                       1310
TVs                           1255
Name: category, dtype: int64

In [13]:
all_items = set(train_df.category.values)
all_items_dict = {k:v for v,k in enumerate(all_items)}
all_items_dict

{'Back to School': 16,
 'Beauty Products': 17,
 'Board Games': 9,
 'Books': 11,
 'Consumer Durables': 8,
 'Ereaders': 7,
 'Fashion': 15,
 'Fitness': 2,
 'Gaming': 4,
 'Groceries': 13,
 'Home Decor': 1,
 'Kitchen cleaning Supplies': 12,
 'Laptops': 6,
 'Painiting Supplies': 3,
 'Pet Supplies': 10,
 'Phones': 0,
 'TVs': 5,
 'Toys': 14}

In [14]:
len(train_df.user_id.unique())

29972

In [15]:
from sklearn.preprocessing import StandardScaler
scalar = StandardScaler()
aov_data = train_df.aov.to_numpy()
scalar.fit(aov_data.reshape(-1,1))
train_df.aov = scalar.transform(train_df.aov.to_numpy().reshape(-1,1))
train_df.head()

user_id       aov    category
0    37327  2.116032      Phones
1    37327 -0.611963     Fashion
2    37327 -0.601914  Home Decor
3    37327 -0.573377     Fashion
4    41260  0.933781      Phones

Train_df has 29972 user data but train_target_df has only 13245 user data

In [16]:
new_df = pd.DataFrame(data=train_target_df.user_id, columns=[['user_id']])
for items in all_items:
  new_df[items] = np.nan

new_df

Empty DataFrame
Columns: [(user_id,), (Phones,), (Home Decor,), (Fitness,), (Painiting Supplies,), (Gaming,), (TVs,), (Laptops,), (Ereaders,), (Consumer Durables,), (Board Games,), (Pet Supplies,), (Books,), (Kitchen cleaning Supplies,), (Groceries,), (Toys,), (Fashion,), (Back to School,), (Beauty Products,)]
Index: []

In [17]:
row_len = new_df.shape

In [18]:
for i,row in enumerate(train_target_df.iterrows()):
  print(type(row))
  row_val = np.zeros(19)
  row_val[0] = row[1].user_id
  temp_df = train_df[train_df.user_id == row_val[0]]
  print(temp_df)
  for temp_df_row in temp_df.iterrows():
    category = temp_df_row[1].category
    row_val[all_items_dict[category] + 1] += temp_df_row[1].aov #all_item_dict is 0 indexed but here row[0] = user-id

  new_df.loc[i] = row_val

new_df.head()

Streaming output truncated to the last 5000 lines.
54044     1032 -0.593097            Fashion
<class 'tuple'>
       user_id       aov           category
36809     1084  1.349418             Phones
36810     1084  0.078426             Phones
36811     1084  1.126904             Phones
36812     1084 -0.615187         Home Decor
36813     1084 -0.606275            Fashion
36814     1084 -0.059330  Consumer Durables
36815     1084 -0.469278            Fitness
36816     1084 -0.583900     Back to School
36817     1084 -0.607507     Back to School
36818     1084 -0.565697            Fashion
<class 'tuple'>
       user_id       aov    category
58106     1209  0.029695      Phones
58107     1209  1.526044      Phones
58108     1209 -0.322136     Fitness
58109     1209 -0.612627  Home Decor
58110     1209 -0.587029     Fashion
58111     1209 -0.614049  Home Decor
<class 'tuple'>
        user_id       aov   category
140006     1231  0.210683    Fitness
140007     1231 -0.591769      Books
140

user_id    Phones Home Decor  ...   Fashion Back to School Beauty Products
0  13153.0  0.624633  -0.605706  ... -0.587787      -0.596984        0.000000
1  42853.0  1.338515  -0.602388  ... -0.591485       0.000000        0.000000
2  30550.0  1.128231  -0.619832  ... -2.318536       0.000000       -0.607128
3   9797.0  0.660755   0.000000  ... -1.167516      -0.610826        0.000000
4   8171.0  2.283219  -0.598880  ... -2.325836       0.000000        0.000000

[5 rows x 19 columns]

In [19]:
new_df.shape

(13245, 19)

In [20]:
#label encoding, let's see
new_labels = train_target_df.category
print(type(new_labels))
new_labels = new_labels.map(all_items_dict)
new_labels

<class 'pandas.core.series.Series'>


0         0
1         0
2         0
3         0
4         0
         ..
13240    15
13241     1
13242     2
13243     2
13244    15
Name: category, Length: 13245, dtype: int64

In [21]:
X, y = new_df.drop(columns=['user_id'], axis=1), new_labels
from xgboost import XGBClassifier
clf = XGBClassifier()
clf.fit(X,y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [22]:
clf.score(X,y)

0.8757266893167233

---

In [23]:
test_df.head()

Unnamed: 0  user_id      aov category
0           0    43323  19115.0   Phones
1           1    43323  29309.0   Phones
2           2    43323  15293.0   Phones
3           3    43323  23548.0   Phones
4           4    43323    751.0  Fashion

In [24]:
test_df.drop(columns=['Unnamed: 0'], inplace=True, axis=1)
test_df.head()

user_id      aov category
0    43323  19115.0   Phones
1    43323  29309.0   Phones
2    43323  15293.0   Phones
3    43323  23548.0   Phones
4    43323    751.0  Fashion

In [25]:
test_df.aov = scalar.transform(test_df.aov.to_numpy().reshape(-1,1))
test_df.head()

user_id       aov category
0    43323  1.166723   Phones
1    43323  2.133193   Phones
2    43323  0.804368   Phones
3    43323  1.587006   Phones
4    43323 -0.574325  Fashion

In [26]:
len(test_df.user_id.unique())

19981

In [27]:
new_test_df = pd.DataFrame(data=None, columns=new_df.columns,)
new_test_df

Empty DataFrame
Columns: [(user_id,), (Phones,), (Home Decor,), (Fitness,), (Painiting Supplies,), (Gaming,), (TVs,), (Laptops,), (Ereaders,), (Consumer Durables,), (Board Games,), (Pet Supplies,), (Books,), (Kitchen cleaning Supplies,), (Groceries,), (Toys,), (Fashion,), (Back to School,), (Beauty Products,)]
Index: []

In [28]:
test_user_ids = test_df.user_id.unique()
for i,uid in enumerate(test_user_ids):
  #print(type(row))
  row_val = np.zeros(19)
  row_val[0] = uid
  temp_df = test_df[test_df.user_id == row_val[0]]
  #print(temp_df)
  for temp_df_row in temp_df.iterrows():
    category = temp_df_row[1].category
    row_val[all_items_dict[category] + 1] += temp_df_row[1].aov #all_item_dict is 0 indexed but here row[0] = user-id

  new_test_df.loc[i] = row_val

new_test_df.head()

user_id    Phones Home Decor  ...   Fashion Back to School Beauty Products
0  43323.0  5.691290   0.000000  ... -2.873520      -1.207620        0.000000
1   4269.0  1.850002  -0.634717  ... -1.722785      -0.608550        0.000000
2  21731.0  4.087049  -0.604568  ... -1.169412       0.000000        0.000000
3  14422.0  1.412106  -0.604094  ... -1.167042      -0.592054       -0.608076
4  12862.0  1.290812  -1.204491  ... -1.719277       0.000000        0.000000

[5 rows x 19 columns]

In [29]:
sam_sub = pd.read_csv('/content/Sample Submission.csv')
sam_sub.head()

user_id                                pred3
0    43323  Laptops, Fashion, Consumer Durables
1     4269  Laptops, Fashion, Consumer Durables
2    21731  Laptops, Fashion, Consumer Durables
3    14422  Laptops, Fashion, Consumer Durables
4    12862  Laptops, Fashion, Consumer Durables

In [30]:
pred = clf.predict_proba(new_test_df.drop(columns=['user_id'], axis=1))

In [48]:
pred_top_3 = []
for pre in pred:
  temp_arr = []
  pre = pre.tolist()
  for i in range(3):
    index = pre.index(max(pre))
    pre[index] = -1
    temp_arr.append(index)

  pred_top_3.append(temp_arr)

In [ ]:
pred_top_3_str = []
inverse_dict = {v:k for k,v in all_items_dict.items()}
for pre in pred_top_3:
  s = inverse_dict[pre[0]] + ', ' + inverse_dict[pre[1]] + ', ' + inverse_dict[pre[2]]
  pred_top_3_str.append(s)

#pred_top_3_str

In [43]:
inverse_dict

{0: 'Phones',
 1: 'Home Decor',
 2: 'Fitness',
 3: 'Painiting Supplies',
 4: 'Gaming',
 5: 'TVs',
 6: 'Laptops',
 7: 'Ereaders',
 8: 'Consumer Durables',
 9: 'Board Games',
 10: 'Pet Supplies',
 11: 'Books',
 12: 'Kitchen cleaning Supplies',
 13: 'Groceries',
 14: 'Toys',
 15: 'Fashion',
 16: 'Back to School',
 17: 'Beauty Products'}

In [44]:
all_items_dict

{'Back to School': 16,
 'Beauty Products': 17,
 'Board Games': 9,
 'Books': 11,
 'Consumer Durables': 8,
 'Ereaders': 7,
 'Fashion': 15,
 'Fitness': 2,
 'Gaming': 4,
 'Groceries': 13,
 'Home Decor': 1,
 'Kitchen cleaning Supplies': 12,
 'Laptops': 6,
 'Painiting Supplies': 3,
 'Pet Supplies': 10,
 'Phones': 0,
 'TVs': 5,
 'Toys': 14}

In [50]:
sam_sub['pred3'] = pred_top_3_str
sam_sub.head()

user_id                                          pred3
0    43323  Phones, Kitchen cleaning Supplies, Home Decor
1     4269     Kitchen cleaning Supplies, Laptops, Phones
2    21731  Phones, Kitchen cleaning Supplies, Home Decor
3    14422  Home Decor, Phones, Kitchen cleaning Supplies
4    12862  Phones, Kitchen cleaning Supplies, Home Decor

In [51]:
sam_sub.to_csv('subb.csv', index=False)

---
---

In [ ]:
pred1 = clf.predict(new_test_df.drop(columns=['user_id'], axis=1))
pred1

array([11, 14, 11, ..., 11, 11, 11])

In [ ]:
new_test_df2 = new_test_df
for i, pr in enumerate(pred1):
  new_test_df2.iloc[i][pr+1] = 0

pred2 = clf.predict(new_test_df2.drop(columns=['user_id'], axis=1))

new_test_df3 = new_test_df
for i, pr in enumerate(pred2):
  new_test_df3.iloc[i][pr+1] = 0

pred3 = clf.predict(new_test_df3.drop(columns=['user_id'], axis=1))

In [ ]:
sam_sub['p1'],sam_sub['p2'],sam_sub['p3'] = pred1,pred2,pred3
sam_sub.head()

user_id                                pred3  p1  p2  p3
0    43323  Laptops, Fashion, Consumer Durables  11  11  11
1     4269  Laptops, Fashion, Consumer Durables  14  14  14
2    21731  Laptops, Fashion, Consumer Durables  11  11  11
3    14422  Laptops, Fashion, Consumer Durables  12  12  12
4    12862  Laptops, Fashion, Consumer Durables  11  11  11

In [ ]:
sam_sub[sam_sub.p1 != sam_sub.p2]

user_id                                pred3  p1  p2  p3
10       11999  Laptops, Fashion, Consumer Durables   7  11  11
11       43072  Laptops, Fashion, Consumer Durables   5  11  11
12       17439  Laptops, Fashion, Consumer Durables   4  11  11
20       17706  Laptops, Fashion, Consumer Durables   4  11  11
28       29011  Laptops, Fashion, Consumer Durables  17  14  14
...        ...                                  ...  ..  ..  ..
19933    11460  Laptops, Fashion, Consumer Durables  17  11  11
19934    32137  Laptops, Fashion, Consumer Durables   7  11  11
19944    12143  Laptops, Fashion, Consumer Durables   7  11  11
19958    22624  Laptops, Fashion, Consumer Durables  17  11  11
19976    42011  Laptops, Fashion, Consumer Durables  17  11  11

[3380 rows x 5 columns]

For now many pred are same but let's see 

In [ ]:
inverse_dict = {v:k for k,v in all_items_dict.items()}
sam_sub.p1 = sam_sub.p1.map(inverse_dict)
sam_sub.p2 = sam_sub.p2.map(inverse_dict)
sam_sub.p3 = sam_sub.p3.map(inverse_dict)
sam_sub

user_id                                pred3  ...          p2          p3
0        43323  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
1         4269  Laptops, Fashion, Consumer Durables  ...     Fashion     Fashion
2        21731  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
3        14422  Laptops, Fashion, Consumer Durables  ...  Home Decor  Home Decor
4        12862  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
...        ...                                  ...  ...         ...         ...
19976    42011  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
19977    16846  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
19978    25357  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
19979    49936  Laptops, Fashion, Consumer Durables  ...      Phones      Phones
19980    50037  Laptops, Fashion, Consumer Durables  ...      Phones      Phones

[19981 rows x 5 columns]

In [ ]:
sam_sub.pred3 = sam_sub.p1 + ', '+ sam_sub.p2 + ', ' + sam_sub.p3
sam_sub.head()

user_id                               pred3  ...          p2          p3
0    43323              Phones, Phones, Phones  ...      Phones      Phones
1     4269           Fashion, Fashion, Fashion  ...     Fashion     Fashion
2    21731              Phones, Phones, Phones  ...      Phones      Phones
3    14422  Home Decor, Home Decor, Home Decor  ...  Home Decor  Home Decor
4    12862              Phones, Phones, Phones  ...      Phones      Phones

[5 rows x 5 columns]

In [ ]:
sam_sub.drop(columns=['p1','p2','p3'], inplace=True, axis=1)
sam_sub.head()

user_id                               pred3
0    43323              Phones, Phones, Phones
1     4269           Fashion, Fashion, Fashion
2    21731              Phones, Phones, Phones
3    14422  Home Decor, Home Decor, Home Decor
4    12862              Phones, Phones, Phones

In [ ]:
sam_sub.to_csv('sub.csv', index=False)